Rewrite `step` and refactor `interpret`

havijw opened the pull request:
I ended up rewriting `step` to do everything on its own without calling `interpret`, and refactored `interpret` a bit with more documentation. `Step` now handles repetition (with/out accumulation) and showing which rules run at each step. `step` definitely got more complicated as a result, but to me it seems preferable to have these functions be separate since they need different things for printing messages along the way and optimizing memoization.

There's plenty of documentation in `step` now, and I added some for `interpret` and some related functions nearby. I did a bit of refactoring of `interpret` — it still does everything the same way, but I think I cleaned things up a little bit. I also slightly changed how the memoization works — now `memo` has nested dicts mapping {input : {binding : (output, rule used)}}, rather than {input : (binding, output, rule used)} like it was before. This is slightly more efficient because it's constant time to look up whether a binding has been seen before, but I doubt this makes a real difference.

I didn't implement memoization for `step`, but I can add it — it'll be a little more complicated than `interpret` but should still be doable. Honestly, I'm not sure it's necessary though, since the amount of output you'd need to be generating for it to make a real difference is massive (at least 9 generations of the MIU system, which is ~220k outputs).

Examples: [test-step.pdf](https://github.com/EasyArray/phosphorus/files/8074586/test-step.pdf)

EasyArray commented:
Thanks! Much needed.

I think I'll hold off on merging this one until HW3 has been submitted, since it actually does use the memoization -- unless you want to (download and) run the Kaggle HW3, solved, against your `step` branch?

Also, I didn't see any of the rules in your pdf attachment, even though showRules=True was included. I do see the logic for showing them in the code, though, so I'll test it out after HW3.

havijw commented:
No idea why the rules weren't shown. Must be an error in the pdf conversion, because they showed up when I actually ran the notebook...

For reference, this is what it looks like: <img width="1362" alt="showing_rules" src="https://user-images.githubusercontent.com/44988777/154154532-8ef2af90-9c19-49bd-9348-20471fef1d24.png">

havijw commented:
I was using homework 3 as part of my testing while I wrote this. I can run through everything to make sure it works properly!

Also, does the version of phosphorus that you have on Kaggle stay up to date with the version on github? I know it's not redownloading every time any more so I wasn't sure.

EasyArray commented:
No, if you are testing *on Kaggle*, you'll have to update the `phosphorus` data source. We *can* set it to update automatically from GitHub, but right now I have to update it manually.

havijw commented:
Just ran through a completed version of HW3 against this branch. Everything ran fine, with negligible differences in runtime vs the main branch! I'm happy to leave it unmerged, just glad I didn't anything broken.

I can figure out a way to send the notebook if you'd like to see it. They can't be attached to pull requests like I had hoped, but I can make a pdf/html version to upload.

havijw commented:
Just did regression tests using every notebook we've had so far. I made a couple small changes — caught a bug in `interpret` to do with handling `showparse`, which didn't show up in HW3 but did in HW 1 and 2; and changed `step` to default to having accumulation off, since that matches how `repeat` works in the early notebooks. I also removed `nstep` since it has no purpose any more and didn't show up in any of the notebooks.

Also just as a note — you actually included a way to make trees for formal systems in the MIU and PQ notebooks! So it shouldn't be too hard to add to `step`, but it also works well as its own thing.

I'm confident everything works as expected now!

EasyArray reviewed (changes requested):
Just getting back to this. It looks good, but now it conflicts with some changes I made. Could you figure out how to properly merge them? @havijw

havijw commented:
I fixed the merge; this version of `interpret` didn't have the `raise_errors` parameter so it was a quick fix.
